### PR TITLE
XAML warnings/diagnostics support

### LINKED ISSUE #5842: Add msbuild warning messages when compile xaml
**Is your feature request related to a problem? Please describe.**
Add msbuild warning messages when compile xaml

image

**Describe the solution you'd like**
During Build, I want to see the message warning, like here

image

### LINKED ISSUE #7461: StyleInclude inside of MergedDictionaries shouldn't be allowed
**Describe the issue**
Right now it's possible to write code below, compiler won't trigger any errors
```xaml
    <Window.Resources>
        <ResourceDictionary>
            <ResourceDictionary.MergedDictionaries>
                <StyleInclude Source="avares://AvaloniaTest.Styles/Generic.axaml" />
            </ResourceDictionary.MergedDictionaries>
        </ResourceDictionary>
    </Window.Resources>
```
It is allowed only because StyleInclude implements IResourceProvider and inner styles might include any resources.
But developers can expect that it also will include inner Styles, which isn't how it actually work (neither how it supposed to work).

### LINKED ISSUE #9612: Make "Duplicate setter" error show at design/preview time
**Is your feature request related to a problem? Please describe.**

With the recent changes for styling, an exception is now thrown if duplicate setters are placed in the same Style/ControlTheme block:
```Xaml
<!-- This was previously swallowed, now throws a runtime exception -->
<Style Selector="Control">
    <Setter Property="Height" Value="20" />
    <Setter Property="Height" Value="32" />
</Style>
```

The problem is this exception is thrown only at runtime, which is really inconvenient, and it really should be caught in the designer when the Xaml is parsed. 

As a side note, +1 for the error message with this though. Including the full Selector (including where nested) was really helpful to find out where I was mistakenly doing this.

### PR DESCRIPTION
**Pull request by maxkatz6**
## What does the pull request do?

1. Support for multiple reported errors. Previously single binding/markupextension/other error would stop whole build at once. Note, XML parsing errors are still fatal for the build. As well as Emit errors. It's planned to support multiple XML parsing errors though, in the future.
2. It is possible to report warnings that can be collected by the IDE or runtime xaml loader (via `RuntimeXamlLoaderConfiguration.DiagnosticHandler` API).
3. It is possible to override diagnostic severity using `TreatWarningsAsErrors`, `WarningsAsErrors`, 
`WarningsNotAsErrors`, `NoWarn`, `EditorConfig` files (limited). Or in case of runtime loader - using the same `RuntimeXamlLoaderConfiguration.DiagnosticHandler`. Note, each diagnostic has default and min severity - you can't mark any Error as a Warning, but you can do in the opposite way.
4. Special handling for Bindings, Selectors/Setters and intrinsic (both XAML standard like x:Static and Avalonia parsers).
5. Added Obsolete warnings and errors support.
6. Standartized avalonia diagnostic codes for a bit.
7. Added "AvaloniaXamlVerboseExceptions" msbuild property. Previously any unhandled exception would log full exception info with stracktrace. Now it's disabled by default.

## What is the current behavior?

Poor.

## How was the solution implemented (if it's not obvious)?

There are a couple of important moments:
1. Compiler and transformation context now have a diagnostics handler built in, which can be used to manually report specific errors/warnings and recover gracefully. 
2. Throwing an exception from the transformer is still supported and is still used almost everywhere. When any transformer fails, visitor will return ISkipXamlAstNode, marking the node to be ignored from further processing, so we won't multiply errors count.
3. Compiler tries to avoid unnecessary errors caused by not crashing early, but there more work needs to be done to perfect this handling, so it is still possible to have a butterfly effect of unhandled errors.
4. In MSBuild Task there is a limit of 100 diagnostics of any type per file.

## Checklist

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [x] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation

## Breaking changes

Different diagnostic codes, but it's hardly a breaking change, as it wasn't really structured.

## Fixed issues

Fixes #5842
Fixes #7461
Re-fixes #9612 (as a warning instead of an error)
Partially #4466

### CODE REVIEW

**Comment by maxkatz6** [comment]
Depends on https://github.com/kekekeks/XamlX/pull/99

File: src/Avalonia.Build.Tasks/XamlCompilerTaskExecutor.cs [hunk 1]
```diff
@@ -376,29 +384,27 @@ bool CompileGroup(IResourceGroup group)
...
-                        int lineNumber = 0, linePosition = 0;
-                        if (e is XamlParseException xe)
-                        {
-                            lineNumber = xe.LineNumber;
-                            linePosition = xe.LinePosition;
-                        }
-
-                        engine.LogError(BuildEngineErrorCode.TransformError, res.FilePath, e, lineNumber, linePosition);
-                        return false;
+                        skipGroupTransformers = true;
```
Comment (maxkatz6) [inline src/Avalonia.Build.Tasks/XamlCompilerTaskExecutor.cs:387]: todo: revert this changes, missing line information and early exit

File: src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/AvaloniaXamlDiagnosticCodes.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
+﻿using System;
+using XamlX;
+
+namespace Avalonia.Markup.Xaml.XamlIl.CompilerExtensions;
+
+internal static class AvaloniaXamlDiagnosticCodes
```
Comment (maxkatz6) [inline src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/AvaloniaXamlDiagnosticCodes.cs:6]: Todo: revisit these values, some of them are unused
Do we need to have unique diagnostic codes per each error? Currently, it's grouped by area and ability to lower severity, i.e. two different selector parse errors will have the same code - AVLN2201 - right now.

File: src/Avalonia.Build.Tasks/XamlCompilerDiagnosticsFilter.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+public class XamlCompilerDiagnosticsFilter
+{
+    private static readonly Regex s_editorConfigRegex =
+        new("""avalonia_xaml_diagnostic\.([\w\d]+)\.severity\s*=\s*(\w*)""");
+
+    private readonly bool _treatWarningsAsErrors;
+    private readonly HashSet<string> _warningsAsErrors;
+    private readonly HashSet<string> _warningsNotAsErrors;
+    private readonly HashSet<string> _noWarn;
```
Comment (maxkatz6) [inline src/Avalonia.Build.Tasks/XamlCompilerDiagnosticsFilter.cs:19]: It works a bit weird, but it seems like we don't need to explicitly handle these properties. And MSBuild will automatically treat warnings as it decides, assuming a diagnostic code is matching.

But still, we need to manually check for editor config, as MSBuild doesn't do it.

Comment (maxkatz6) [reply src/Avalonia.Build.Tasks/XamlCompilerDiagnosticsFilter.cs:19]: todo: remove everything but `analyzerConfigFiles` checks from here.

File: src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/Transformers/AvaloniaXamlIlStyleValidatorTransformer.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+              && context.GetAvaloniaTypes().IStyle.IsAssignableFrom(on.Type.GetClrType())))
+            return node;
+
+        if (context.ParentNodes().FirstOrDefault() is XamlAstXamlPropertyValueNode propertyValueNode
+            && propertyValueNode.Property.GetClrProperty() is { } clrProperty
+            && clrProperty.Name == "MergedDictionaries"
+            && clrProperty.DeclaringType == context.GetAvaloniaTypes().ResourceDictionary)
+        {
+            var nodeName = on.Type.GetClrType().Name;
+            context.ReportDiagnostic(new XamlDiagnostic(
```
Comment (maxkatz6) [inline src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/Transformers/AvaloniaXamlIlStyleValidatorTransformer.cs:22]: Example of Avalonia-specific XAML warning.

**Comment by maxkatz6** [comment]
If anybody has some time to test this PR, it would help. Basically, it should be able to report most XAML errors, while not regressing anything.

**Comment by MrJul** [comment]
I tested this for a while and can see multiple errors being properly reported and MSBuild error properties being respected.
This seems to work nicely!

Here's a few things I've noticed:

## Missing error location

The file name and position where an error occurs don't seem to be properly reported anymore:

Before this PR:

> 1>C:\Dev\Avalonia\samples\Sandbox\MainWindow.axaml(8,6,8,6): Avalonia error AVLN: 0004: XamlX.XamlParseException: Duplicate setter encountered for property 'Height' Line 8, position 6. [C:\Dev\Avalonia\samples\Sandbox\Sandbox.csproj]
     1>C:\Dev\Avalonia\samples\Sandbox\MainWindow.axaml(8,6,8,6): Avalonia error AVLN: 0004:    at Avalonia.Markup.Xaml.XamlIl.CompilerExtensions.Transformers.AvaloniaXamlIlDuplicateSettersChecker.Transform(AstTransformationContext context, IXamlAstNode node) in C:\Dev\Avalonia\src\Markup\Avalonia.Markup.Xaml.Loader\CompilerExtensions\Transformers\AvaloniaXamlIlDuplicateSettersChecker.cs:line 39 [C:\Dev\Avalonia\samples\Sandbox\Sandbox.csproj]
     1>C:\Dev\Avalonia\samples\Sandbox\MainWindow.axaml(8,6,8,6): Avalonia error AVLN: 0004:    at XamlX.Transform.AstTransformationContext.Visitor.Visit(IXamlAstNode node) in C:\Dev\Avalonia\src\Markup\Avalonia.Markup.Xaml.Loader\xamlil.github\src\XamlX\Transform\AstTransformationContext.cs:line 58 [C:\Dev\Avalonia\samples\Sandbox\Sandbox.csproj]
	 [...]
	 
With Rider/VS properly taking me to the error location.

After this PR:

> 28>MSBUILD : Avalonia error AVLN2203: Duplicate setter encountered for property 'Height' [C:\Dev\Avalonia\samples\Sandbox\Sandbox.csproj]

Notice the lack of `MainWindow.axaml(8,6,8,6)`

## Multiple obsolete diagnostics

An obsolete usage is reported as two warnings/errors (probably because it's two lines):

> 23>MSBUILD: Error AVLN5001 Avalonia: 'SomeObsoleteControl' is obsolete [C:\Dev\Avalonia\samples\Sandbox\Sandbox.csproj]
> 23>MSBUILD: Error AVLN5001 Avalonia: don't use me! [C:\Dev\Avalonia\samples\Sandbox\Sandbox.csproj]

In contrast, the C# compiler reports:

> Warning CS0618 : 'SomeObsoleteControl' is obsolete: 'don't use me!'

## Duplicate unexpected errors

Note that while known errors don't have a strack trace anymore, unexpected ones still do, effectively reporting each line as a different error.
For example:

>21>MSBUILD: Error AVLN0000 Avalonia: The input string 'mm' was not in a correct format. [C:\Dev\Avalonia\samples\Sandbox\Sandbox.csproj]
21>MSBUILD: Error AVLN0000 Avalonia: System.FormatException: The input string 'mm' was not in a correct format. [C:\Dev\Avalonia\samples\Sandbox\Sandbox.csproj]
21>MSBUILD: Error AVLN0000 Avalonia: at System.Number.ThrowOverflowOrFormatException(ParsingStatus status, ReadOnlySpan`1 value, TypeCode type) [C:\Dev\Avalonia\samples\Sandbox\Sandbox.csproj]
21>MSBUILD: Error AVLN0000 Avalonia: at System.Double.Parse(String s, IFormatProvider provider) [C:\Dev\Avalonia\samples\Sandbox\Sandbox.csproj]
[...]

However that was already the case before this PR, so it' just a thing to note for later fixing.

## EditorConfig handling

As noted in the code, EditorConfig parsing is very basic, using regexes, so specified file globs are ignored.
Could we use https://github.com/editorconfig/editorconfig-core-net/ instead to properly parse them, and match the globs?

Note that this library targets .NET Standard 1.3, doesn't have any dependency, and is MIT-licensed, so we should be able to ILMerge into Avalonia.Build.Tasks without issues.

**Comment by maxkatz6** [comment]
> Missing error location

Interesting. Will take a look. It should be reported correctly.

> Multiple obsolete diagnostics

Looks like MSBuild doesn't handle multiline errors as an error with description. Instead it splits it into multiple errors/warnings. Need to research on how to correctly add description to the error/warning.

> Duplicate unexpected errors

In general, we can't completely avoid "butterfly effect" of one error producing more errors. But in your example I see something that should be avoided (basically the same parsing error reported multiple time) + AVLN0000 code is wrong there. Will check it.

> EditorConfig handling

Yes, I will try to ilmerge it.

**Comment by workgroupengineering** [comment]
Great work.

A feature I would like to have would be the ability to write an analyzer like the Roslyn analyzers, which can be plugged into the compilation pipeline via nuget.

Everyone could write their own rules based on company policy.

**Comment by maxkatz6** [comment]
@workgroupengineering I don't think we will have plugin/analyzer infrastructure soon. We don't even have an immutable nodes tree that we can safely send to the third-party analyzer. So, for now, the best bet would be to implement these analyzers as a part of the compiler.

**Comment by maxkatz6** [comment]
@MrJul I am going to leave EditorConfig improvements out of the scope of this PR for now. Other issues should be solved.

**Comment by MrJul** [comment]
Tested again, everything works fine (with `AvaloniaXamlVerboseExceptions=false` for the last issue).

Let's get back to the EditorConfig issue later, once your PR there is hopefully merged :)

File: tests/Avalonia.Markup.Xaml.UnitTests/Xaml/StyleIncludeTests.cs [hunk 5]
```diff
@@ -277,6 +277,34 @@ public void StyleInclude_Should_Be_Replaced_With_Direct_Call()
         Assert.IsType<SimpleTheme>(control.Styles[0]);
         Assert.IsType<SimpleTheme>(control.Styles[1]);
     }
+    
+    [Fact]
+    public void StyleInclude_Inside_Resources_Should_Produce_Warning()
```
Comment (MrJul) [inline tests/Avalonia.Markup.Xaml.UnitTests/Xaml/StyleIncludeTests.cs:282]: The test name specifies StyleInclude but there's none in the XAML below?

Comment (maxkatz6) [inline tests/Avalonia.Markup.Xaml.UnitTests/Xaml/StyleIncludeTests.cs:282]:
```suggestion
    public void Style_Inside_Resources_Should_Produce_Warning()
```

**Review by MrJul — APPROVED** [review]
